File: backend/app/schemas/delivery.py
from datetime import date, datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator


ALLOWED_SCHEDULE_STATUSES = {"agendado", "cancelado", "faltou", "reagendado"}
ALLOWED_DELIVERY_STATUSES = {"concluida"}

DeliveryOperationsPeriod = Literal["hoje", "amanha", "semana", "todos"]
DeliveryOperationsStatus = Literal[
    "agendado",
    "reagendado",
    "retirado",
    "ocorrencia",
]


class DeliveryScheduleCreate(BaseModel):
    """Payload para criação de agendamento de entrega."""

    family_id: int
    basket_type_id: int
    scheduled_date: date
    status: str = "agendado"
    notes: str | None = None

    @field_validator("status")
    @classmethod
    def validate_status(cls, value: str) -> str:
        value = value.strip().lower()
        if value not in ALLOWED_SCHEDULE_STATUSES:
            raise ValueError("Status de agendamento inválido.")
        return value


class DeliveryScheduleResponse(BaseModel):
    """Resposta serializada de agendamento de entrega."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    family_id: int
    basket_type_id: int
    scheduled_date: date
    status: str
    notes: str | None
    created_by_user_id: int


class DeliveryOperationItemResponse(BaseModel):
    """Projeção operacional enriquecida para a agenda de entregas."""

    id: int
    family_id: int
    family_code: str
    family_status: str
    basket_type_id: int
    basket_type_name: str
    scheduled_date: date
    status: str
    notes: str | None
    street: str
    number: str
    complement: str | None
    neighborhood: str
    city: str
    state: str


class DeliveryOperationsSummaryResponse(BaseModel):
    """Contadores globais do período, independentes da página atual."""

    scheduled: int
    rescheduled: int
    completed: int
    exceptions: int
    total: int


class DeliveryOperationsResponse(BaseModel):
    items: list[DeliveryOperationItemResponse]
    total: int
    limit: int
    offset: int
    reference_date: date
    period: DeliveryOperationsPeriod
    summary: DeliveryOperationsSummaryResponse


class DeliveryScheduleUpdate(BaseModel):
    scheduled_date: date
    status: str
    notes: str | None = None

    @field_validator("status")
    @classmethod
    def validate_status(cls, value: str) -> str:
        value = value.strip().lower()
        if value not in ALLOWED_SCHEDULE_STATUSES:
            raise ValueError("Status de agendamento invalido.")
        return value


class DeliveryFromScheduleCreate(BaseModel):
    """Payload para confirmar entrega a partir de um agendamento."""

    delivery_date: datetime
    status: str = "concluida"
    notes: str | None = None

    @field_validator("status")
    @classmethod
    def validate_status(cls, value: str) -> str:
        value = value.strip().lower()
        if value not in ALLOWED_DELIVERY_STATUSES:
            raise ValueError("Status de entrega inválido.")
        return value


class DeliveryTraceItemResponse(BaseModel):
    """Item e lote efetivamente consumidos por uma entrega."""

    movement_id: int
    item_id: int
    item_name: str
    unit_measure: str
    batch_id: int
    batch_code: str | None
    batch_status: str
    storage_location: str | None
    expiration_date: date | None
    quantity: int


class DeliveryResponse(BaseModel):
    """Resposta serializada de entrega."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    delivery_schedule_id: int | None
    family_id: int
    basket_type_id: int
    delivery_date: datetime
    delivered_by_user_id: int
    status: str
    notes: str | None
    items: list[DeliveryTraceItemResponse] = Field(default_factory=list)

File: frontend/src/types/delivery.ts
export interface DeliveryScheduleResponse {
  id: number;
  family_id: number;
  basket_type_id: number;
  scheduled_date: string;
  status: string;
  notes: string | null;
  created_by_user_id: number;
}

export type DeliveryOperationsPeriod = "hoje" | "amanha" | "semana" | "todos";

export type DeliveryOperationsStatus =
  | "agendado"
  | "reagendado"
  | "retirado"
  | "ocorrencia";

export interface DeliveryOperationItemResponse {
  id: number;
  family_id: number;
  family_code: string;
  family_status: string;
  basket_type_id: number;
  basket_type_name: string;
  scheduled_date: string;
  status: string;
  notes: string | null;
  street: string;
  number: string;
  complement: string | null;
  neighborhood: string;
  city: string;
  state: string;
}

export interface DeliveryOperationsSummaryResponse {
  scheduled: number;
  rescheduled: number;
  completed: number;
  exceptions: number;
  total: number;
}

export interface DeliveryOperationsResponse {
  items: DeliveryOperationItemResponse[];
  total: number;
  limit: number;
  offset: number;
  reference_date: string;
  period: DeliveryOperationsPeriod;
  summary: DeliveryOperationsSummaryResponse;
}

export interface DeliveryResponse {
  id: number;
  delivery_schedule_id: number | null;
  family_id: number;
  basket_type_id: number;
  delivery_date: string;
  delivered_by_user_id: number;
  status: string;
  notes: string | null;
  items: DeliveryTraceItemResponse[];
}

export interface DeliveryScheduleCreatePayload {
  family_id: number;
  basket_type_id: number;
  scheduled_date: string;
  status: string;
  notes: string | null;
}

export interface DeliveryTraceItemResponse {
  movement_id: number;
  item_id: number;
  item_name: string;
  unit_measure: string;
  batch_id: number;
  batch_code: string | null;
  batch_status: string;
  storage_location: string | null;
  expiration_date: string | null;
  quantity: number;
}

export interface DeliveryScheduleUpdatePayload {
  scheduled_date: string;
  status: string;
  notes: string | null;
}

export interface DeliveryFromScheduleCreatePayload {
  delivery_date: string;
  status: string;
  notes: string | null;
}
